File: src/main/java/maestro/cache/WorldData.java
package maestro.cache;

import java.nio.file.Path;
import maestro.Maestro;
import maestro.api.cache.ICachedWorld;
import maestro.api.cache.IWaypointCollection;
import maestro.api.cache.IWorldData;
import net.minecraft.world.level.dimension.DimensionType;

/**
 * Data about a world, from maestro's point of view. Includes cached chunks, waypoints, and map
 * data.
 *
 * @author leijurv
 */
public class WorldData implements IWorldData {

    public final CachedWorld cache;
    private final WaypointCollection waypoints;
    // public final MapData map;
    public final Path directory;
    public final DimensionType dimension;

    WorldData(Path directory, DimensionType dimension) {
        this.directory = directory;
        this.cache = new CachedWorld(directory.resolve("cache"), dimension);
        this.waypoints = new WaypointCollection(directory.resolve("waypoints"));
        this.dimension = dimension;
    }

    public void onClose() {
        Maestro.getExecutor()
                .execute(
                        () -> {
                            System.out.println("Started saving the world in a new thread");
                            cache.save();
                        });
    }

    @Override
    public ICachedWorld getCachedWorld() {
        return this.cache;
    }

    @Override
    public IWaypointCollection getWaypoints() {
        return this.waypoints;
    }
}

File: common/src/main/kotlin/maestro/cache/WorldData.kt
package maestro.cache

import maestro.Agent
import maestro.api.cache.ICachedWorld
import maestro.api.cache.IWaypointCollection
import maestro.api.cache.IWorldData
import maestro.api.utils.MaestroLogger
import net.minecraft.world.level.dimension.DimensionType
import java.nio.file.Path

private val log = MaestroLogger.get("cache")

/**
 * Data about a world, from maestro's point of view. Includes cached chunks, waypoints, and map
 * data.
 */
class WorldData internal constructor(
    val directory: Path,
    val dimension: DimensionType,
) : IWorldData {
    val cache: CachedWorld = CachedWorld(directory.resolve("cache"), dimension)
    private val waypoints: WaypointCollection = WaypointCollection(directory.resolve("waypoints"))

    fun onClose() {
        Agent.getExecutor().execute {
            log.atInfo().log("World save started")
            cache.save()
        }
    }

    override fun getCachedWorld(): ICachedWorld = cache

    override fun getWaypoints(): IWaypointCollection = waypoints
}
